devlab/tests/testcases/test_resource_migration.py

```diff
@@ -334,6 +334,39 @@ def test_migrate_glance_images(self):
                                                 resource_name='image',
                                                 parameter='id')
 
+    @attr(migrated_tenant=['admin', 'tenant1', 'tenant2'])
+    def test_migrate_deleted_glance_images_only_once(self):
+        """Validate deleted and broken images were migrated to dst only once.
+        """
+        src_vms = self.src_cloud.novaclient.servers.list(
+            search_opts={'all_tenants': 1})
+        src_img_ids = [i.id for i in self.src_cloud.glanceclient.images.list()]
+        # getting images, from which vms were spawned, but which do not exist
+        # in the glance
+        to_restore_img_ids = []
+        for vm in src_vms:
+            if vm.image and vm.image['id'] not in src_img_ids:
+                to_restore_img_ids.append(vm.image['id'])
+        # getting 'broken' images (which exist in the glance, but deleted in
+        # storage)
+        all_images = self.migration_utils.get_all_images_from_config()
+        broken_images = [i['name'] for i in all_images if i.get('broken')]
+        src_images = self.src_cloud.glanceclient.images.list()
+        to_restore_img_ids.extend([image.id for image in src_images
+                                  if image.name in broken_images])
+
+        dst_images = [x for x in self.dst_cloud.glanceclient.images.list()]
+        restored_dst_images = []
+        for deleted_img_id in set(to_restore_img_ids):
+            for dst_image in dst_images:
+                if dst_image.name and deleted_img_id in dst_image.name:
+                    restored_dst_images.append(dst_image.name)
+        duplicated = [i for i in restored_dst_images
+                      if restored_dst_images.count(i) > 1]
+        if duplicated:
+            msg = 'Following images were restored more than 1 time: "%s"'
```

Comment:
Would be more readable with number of actual restores, e.g. "Image X restored N times"